src/metric.h

```diff
@@ -0,0 +1,177 @@
+// Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <string>
+#include "ipc_message.h"
+#include "pb_string.h"
+#include "pb_utils.h"
+
+#ifdef TRITON_PB_STUB
+#include <pybind11/embed.h>
+namespace py = pybind11;
+#else
+#include "triton/core/tritonserver.h"
+#endif
+
+namespace triton { namespace backend { namespace python {
+
+enum MetricRequestKind {
+  MetricNew,
+  MetricDelete,
+  MetricValue,
+  MetricIncrement,
+  MetricSet
+};
+
+struct MetricShm {
+  bi::managed_external_buffer::handle_t family_name_shm_handle;
+  bi::managed_external_buffer::handle_t labels_shm_handle;
+  double value;
+  MetricRequestKind metric_request_kind;
+  void* metric_address;
+  void* metric_family_address;
+};
+
+class Metric {
+ public:
+  Metric(
+      const std::string& family_name, const std::string& labels,
+      void* metric_family_address);
+
+  ~Metric();
+
+  /// Get the family name of the metric.
+  /// \return Returns the family name of the metric.
+  const std::string& FamilyName();
+
+  /// Get the labels of the metric.
+  /// \return Returns the labels of the metric.
+  const std::string& Labels();
+
+  /// Get the shared memory handle of the metric.
+  /// \return Returns the shared memory handle of the metric.
+  bi::managed_external_buffer::handle_t ShmHandle();
+
+  /// Get the request kind of the metric.
+  /// \return Returns the request kind of the metric.
+  const MetricRequestKind& RequestKind();
```

Comment:
I'm not sure I understand why the Metric/MetricFamily objects have a RequestKind(). Shouldn't the same long-lived metric/metric family object be created, incremented, set, deleted, etc.? What does it mean for the object to have a RequestKind? Isn't that a moreso a property of the ShmHandles passed around for communication?

Reply:
Removed all the request kind and use `PYTHONSTUB_CommandType` in `IPCMessage` instead.